Avx debug

sarahsummerfield opened the pull request:
No longer segfaulting and test files have been standardized. All that's left is tweaking the actual calculation stuff to get the right numbers. I think this no longer constitutes an actual "debug", so it would be a good time to merge this back into the AVX feature branch.

sarahsummerfield committed:
extraction stage changes finished. Result not yet confirmed, but it is able to compile now!

sarahsummerfield committed:
Still debugging, this is just a safety push

sarahsummerfield committed:
Fix initial segfaulting and test files

Changes made in TreeBuilderManager.cpp to fix probable memory alignment issues with the Distance Calculator. This fixed the initial segfaulting in NewCalcDNA (but there is more to figure out.)
Also made changes to the fixtures - random.faa changed to amino.faa to indicate it has amino acid sequences in it. Generated and added DNA.fa, for testing the DNA functions we've been working on.
Program is currently producing a tree with amino.faa but segfaulting at src/DistanceCalculatorAVX.cpp:368 with the dna file.

sarahsummerfield committed:
Fix all segfaults and produce a DNA tree

Made two major changes that fixed the segfaulting issue on DNA
sequences and allow it to produce an (albeit not quite correct still)
tree.
NewCalcDNA now explicitly loads a chunk from the character arrays into
seq1/seq2. Was able to change it from being an unaligned load by
changing the allocation of convertedSequences in ConvertAllDNA to use
aligned_alloc instead of new.

glesica commented on src/DistanceCalculatorAVX.cpp:
```diff
@@ -364,11 +365,16 @@ double DistanceCalculator::newCalcDNA(int a, int b) {
-            gap1 = *(__m256i *)&Agap[i];
-            gap2 = *(__m256i *)&Bgap[i];
+            seq1 = _mm256_load_si256((__m256i*)&Achar[i]);
+            //seq1 = *(__m256i *)&Achar[i];
```
Why'd you leave these commented? Can we just delete them?

sarahsummerfield replied:
I didn't have a firm reason, other than I suspect we're going to have to make a lot of similar changes through the codebase and I find it helpful having the before and after when comparing to other areas... we don't need to keep them in, though!

glesica commented on src/DistanceCalculatorAVX.cpp:
```diff
@@ -377,45 +383,57 @@ double DistanceCalculator::newCalcDNA(int a, int b) {
+        consecutive 8 differences to produce four unsigned 16-bit
+        integers, and pack these unsigned 16-bit integers in the
+        low 16 bits of 64-bit elements in dst.
+        **in avx this gives 4 16-bit integers, not 2*/
```
This comment is a little confusing. We're packing four 16-bit integers into a single 64-bit element? Or one per 64-bit element?

sarahsummerfield replied:
It's packing 1 integer into the low 16 bits of each of the 64 bit blocks... Slightly confusing phrasing courtesy of the Intel Intrinsics guide.

glesica commented on src/TreeBuilderManager.cpp:
```diff
@@ -99,10 +99,12 @@ std::string TreeBuilderManager::doJob() {
+            new(distCalc) DistanceCalculator(
+                    seqs, (DistanceCalculator::AlphabetType)alphType,
+                    (DistanceCalculator::CorrectionType)corrType,
+                    seqReader->numSeqs, this->newDistanceMethod);
```
Do we still need to do this?

sarahsummerfield replied:
I think we do, until/if we change all the field allocations to use aligned_alloc.

sarahsummerfield committed:
Skip clang-format test

Clang-format seems to be having issues, we're skipping those tests for
now so the relevant lines in the makefile have been commented out.

sarahsummerfield committed:
Complete the makefile

Somehow a large chunk of the makefile that exists in the AVX branch was
missing from this debug branch. This makefile has been updated to add
the missing parts, with check-format still being skipped.

sarahsummerfield committed:
Merge branch 'avx' into avx-debug

sarahsummerfield committed:
Fix makefile echo issue

Small formatting issue in check-format echo statement fixed so test
will actually be skipped now... for real.

sarahsummerfield committed:
Merge branch 'avx-debug' of https://github.com/TravisWheelerLab/NINJA into avx-debug

I'm not sure why I have to pull, but I'm listening to git...

sarahsummerfield committed:
Make modifications for PR feedback

Old, commented out seq/gap initialization marked as previous versions
Comment on gather stage clarified to denote separate lanes
Deleting the empty mistake of a test file

sarahsummerfield commented:
@glesica I've pushed the changes that we discussed on Friday!
